src/sisl/io/siesta/vibra.py

```diff
@@ -0,0 +1,257 @@
+        """Open the file
+
+        Here we also initialize some variables to keep track of the state of the read.
+        """
+        super()._open()
+        if rewind:
+            self.seek(0)
+        # Here we initialize the variables that will keep track of the state of the read.
+        # The process for identification is done on this basis:
+        #  _ik is the current (Python) index for the k-point to be read
+        #  _state is:
+        #        -1 : the file-descriptor has just been opened (i.e. in front of header)
+        #         0 : it means that the file-descriptor is in front of k point information
+        self._state = -1
+        self._ik = 0
+
+    def _setup_parsing(self, close=True):
+        """Gets all the things needed to parse the wfsx file.
+
+        Parameters
+        -----------
+        close: bool, optional
+            Whether the file unit should be closed afterwards.
+        """
+        self._open()
+        # Read the sizes relevant to the file.
+        # We also read whether there's only gamma point information or there are multiple points
+        if self._state == -1:
+            self._sizes = self._r_sizes()
+            self._state = 0
+
+        if close:
+            self._close()
+
+    def _r_sizes(self):
+        """ Determine the dimension of the data stored in the vectors file
+        """
+        pos = self.fh.tell()
+        # Skip past header
+        self.readline() # Empty line
+        self.readline() # K point
+        self.readline() # Eigenmode index
+        self.readline() # Frequency
+        self.readline() # Eigenmode header
+        
+        # Determine number of atoms (i.e. rows per mode)
+        natoms = 0
+        while True:
+            line = self.readline()
+            if line.startswith('Eigenmode'):
+                break
+            else:
+                natoms += 1
+        self._natoms = natoms
+
+        # Determine number of modes
+        nmodes = 1
+        while True:
+            line = self.readline()
+            if line.startswith('k            ='):
+                break 
+            elif line.startswith('Eigenvector'):
+                nmodes += 1
+        self._nmodes = nmodes
+
+        # Rewind
+        self.seek(pos)
+
+    def _r_next_eigenmode(self):
+        """Reads the next phonon eigenmode in the vectors file.
+
+        Parameters
+        ----------
+        ik: integer
+            the (python) k index of the next eigenstate.
+
+        Returns
+        --------
+        EigenmodePhonon:
+            The next eigenmode.
+        """
+        # Skip empty line at the head of the file
+        self.readline()
+            
+        k = list(map(float, self.readline().split()[2:]))
+        if len(k) == 0:
+            raise GeneratorExit
+        # Read first eigenvector index 
+        
+        # Determine number of atoms (i.e. rows per mode)
+        state = np.empty((self._nmodes, self._natoms, 3), dtype=np.complex128)
```

Comment:
If I am not mistaken, then `state` should have final shape `nmodes, natoms * 3`, not a 3D array.

Reply:
Based on the way it is used in `DynamicalMatrix.eigenmode` I agree with you. What I am less sure about is the order should be. Based on the way the dynamical matrix is constructed from FC output, I understand that it should be [u_1x, u_1y, u_1z, u_2x, ...] right?

Reply:
yes, you could create it like this, but when storing, simply do `state.reshape(nmodes, -1)`, then the correct size is passed.